ecies/ecies.go

```diff
@@ -0,0 +1,129 @@
+package ecies
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/hmac"
+	"crypto/rand"
+	"crypto/sha256"
+	"crypto/subtle"
+	"errors"
+
+	"github.com/decred/dcrd/dcrec/secp256k1/v4"
+)
+
+// Encrypts msg to destPubKey using the following construction:
+//
+// r = random number (private key)
+// R = r * G (public key)
+// S = Px where (Px, Py) = r * KB
+// kE || kM = KDF(S, 32)
+// iv = random initialization vector
+// c = c = AES(kE, iv , m)
+// d = MAC(sha256(kM), iv || c)
+// msg = R || iv || c || d
+//
+// For more details, see the ECIES Encryption docs defined by Eth's DevP2P:
+// https://github.com/ethereum/devp2p/blob/master/rlpx.md#ecies-encryption
+func Encrypt(destPubKey *secp256k1.PublicKey, msg []byte) ([]byte, error) {
+	r, err := secp256k1.GeneratePrivateKey()
+	if err != nil {
+		return nil, err
+	}
+
+	var (
+		s  = secp256k1.GenerateSharedSecret(r, destPubKey)
+		k  = kdf(s)
+		ke = k[:16]
+		km = sha256.Sum256(k[16:])
+	)
+
+	iv := make([]byte, aes.BlockSize)
+	rand.Read(iv)
+
+	block, err := aes.NewCipher(ke)
+	if err != nil {
+		return nil, err
+	}
+	c := make([]byte, len(msg))
+	cipher.NewCTR(block, iv).XORKeyStream(c, msg)
+
+	mac := hmac.New(sha256.New, km[:])
+	mac.Write(c)
+	d := mac.Sum(nil)
+
+	var res []byte
+	res = append(res, r.PubKey().SerializeCompressed()...)
+	res = append(res, iv...)
+	res = append(res, c...)
+	res = append(res, d...)
+	return res, nil
+}
+
+// Decrypts ciphertext using prvKey using the following construction:
+//
+// ciphertext = R || iv || c || d
+// S = Px where (Px, Py) = kB * R
+// kE || kM = KDF(S, 32)
+// d == MAC(sha256(kM), iv || c)
+// msg = AES(kE, iv || c)
+func Decrypt(prvKey *secp256k1.PrivateKey, ciphertext []byte) ([]byte, error) {
+	const (
+		pubKeyLen = 33
+		ivLen     = 16
+		macSize   = 32
+	)
+	if len(ciphertext) <= pubKeyLen+ivLen+macSize {
+		return nil, errors.New("ciphertext is too small")
+	}
+	var (
+		msgStart = pubKeyLen + ivLen
+		msgEnd   = len(ciphertext) - macSize
+	)
+
+	r, err := secp256k1.ParsePubKey(ciphertext[:pubKeyLen])
+	if err != nil {
+		return nil, err
+	}
+	var (
+		s  = secp256k1.GenerateSharedSecret(prvKey, r)
+		k  = kdf(s)
+		ke = k[:16]
+		km = sha256.Sum256(k[16:])
+	)
+
+	mac := hmac.New(sha256.New, km[:])
+	mac.Write(ciphertext[msgStart:msgEnd])
+	if subtle.ConstantTimeCompare(ciphertext[msgEnd:], mac.Sum(nil)) != 1 {
+		return nil, errors.New("invalid hmac")
+	}
+
+	block, err := aes.NewCipher(ke)
+	if err != nil {
+		return nil, err
+	}
+	iv := ciphertext[pubKeyLen : pubKeyLen+ivLen]
+	msg := make([]byte, len(ciphertext[msgStart:msgEnd]))
+	cipher.NewCTR(block, iv).XORKeyStream(msg, ciphertext[msgStart:msgEnd])
+	return msg, nil
+}
+
+// Modified NIST SP 800-56: Concatenation Key Derivation Function.
+// This function avoids incrementing the counter, and therefore
+// looping, since:
+// reps = ceil(keydatalen / hashlen)
+// and keydatalen = 32 and hashlen = 32
```

Comment:
```suggestion
// and keydatalen = 32 and hashlen = 64
```
Not to be pedantic, but technically the block size of sha256 is 512 bits or 64 bytes :)

Reply:
First of all, pedantic is what we need! Sweat the small stuff! Prioritize precision!

That being said, while the block size is 512 bits, I think the result of calling sum is 256 bits. In this case, I meant hashlen to be the size of sum. Per the NIST paper, do you read hashlen to be the size of the hash sum or the size of the block?

Reply:
From the paper, "hashlen: an integer equal to the length (in bits) of the output block of the hash function
used to derive blocks of secret keying material"

For [SHA256](https://en.wikipedia.org/wiki/Secure_Hash_Algorithms), the block size is 512 bits whereas each word (and the result) is 256 bits. I think it could mean 512 (or 64 bytes) here, but it doesn't really matter since ceil(32/64) = 1 anyways.